### PR TITLE
feat: Add support for server side Portals

### PR DESCRIPTION
**Pull request by Fryuni**
Working implementation of https://github.com/withastro/roadmap/discussions/1032#discussioncomment-10804001

Still missing the tests to be ready

- `<portal-in to="some name">` elements are voided and their children sent to a matching portal named `"some name"`
- `<portal-out name="some name">` are replaced flatly with the children sent from `<portal to="some name">` (the same effect as `.replaceWith()`)
- Implicit portal at start and end of `<head>`, `<body>` and every element with a target

### CODE REVIEW

**Comment by Fryuni** [comment]
Resulting HTML from changed demo can be seen here:
https://feat-portal-gun.inox-tools-ex-request-nanostores.pages.dev/

image

**Comment by Fryuni** [comment]
> That looks pretty cool!

Thank you, kind sir 😊
It was fun

>     1. I'm not sure using a portal tag is a good idea, given that it [has already usage in HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/portal). 

Yes, I also found that out. Technically, any tag without a `-` is reserved by the HTML spec for future use. That is why custom elements for web components must have a `-`.

Astro already overrides the semantics of `<slot>`, `<script>` and `<style>`. I think doing the same for `<portal>` is fine as long as we include an opt-out in the form of `is:inline` like the others.

> You could maybe provide some components that use data attrs under the hood or something?

The advantage of using a direct HTML tag for this is that it works even when the element was added by an UI framework component.

The disadvantage is that makes such UI framework components don't work correctly on the client side. But as an alternative, users can put the entire UI framework in the portal, which will move the entire Astro Island to the new place and be hydrated there.

>     2. I think this disables streaming? If so, worth mentioning in the docs

Yes, I mentioned that in my comment on the discussion https://github.com/withastro/roadmap/discussions/1032#discussioncomment-10804001
By the concept of how the portal should work I don't think that can be avoided. It will be in the docs.

File: packages/portal-gun/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+Add the integration to your `astro.config.mjs`:
+
+```js
+// astro.config.mjs
+import { defineConfig } from 'astro';
+import portalGun from '@inox-tools/portal-gun';
+
+export default defineConfig({
+  integrations: [portalGun({})],
```
Comment (florian-lefebvre) [inline packages/portal-gun/README.md:23]: Does it need an empty object?

File: packages/portal-gun/package.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "@inox-tools/portal-gun",
+  "version": "0.0.0",
+  "description": "Transport HTML elements through your page during rendering using Portals.",
+  "keywords": [
```
Comment (florian-lefebvre) [inline packages/portal-gun/package.json:5]: I think you may need an integration category

File: packages/portal-gun/package.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    "build": "tsup",
+    "dev": "tsup --watch",
+    "prepublish": "pnpm run build",
+    "test": "vitest run",
+    "test:dev": "vitest"
+  },
+  "dependencies": {
+    "@inox-tools/runtime-logger": "workspace:^",
+    "@inox-tools/utils": "workspace:^",
+    "astro-integration-kit": "catalog:",
```
Comment (florian-lefebvre) [inline packages/portal-gun/package.json:38]: First time I see the catalog in use 👍

File: packages/portal-gun/src/internal/debug.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import debugC from 'debug';
+
+export const debug = debugC('inox-tools:portal-gun')
+
+export const getDebug = (segment?: string) => {
```
Comment (florian-lefebvre) [inline packages/portal-gun/src/internal/debug.ts:5]: What does it do?

File: packages/portal-gun/virtual.d.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	portal: PortalAttrs;
+}
+
+declare namespace JSX {
+	export interface IntrinsicElements extends NewElements {}
+}
+
+import 'preact';
+
+declare module 'preact' {
```
Comment (florian-lefebvre) [inline packages/portal-gun/virtual.d.ts:13]: Why only preact?

File: packages/portal-gun/src/internal/debug.ts (hunk 4 above)
Comment (Fryuni) [reply packages/portal-gun/src/internal/debug.ts:5]: It's part of my template, it isn't being used here.

File: packages/portal-gun/virtual.d.ts (hunk 5 above)
Comment (Fryuni) [reply packages/portal-gun/virtual.d.ts]: I added types to React, Preact, Solid, Svelte and Astro.

**Pull request merged**